client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +155,116 @@ public void start(Listener listener) {
-                        this.bootstrapServers.stream().map(ResolvedServerInfo::new)
-                                .collect(Collectors.toList()))
-                        .build();
-                log.info("Updating client with controllers: {}", serverInfoGroup);
-                this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
+                // If the servers comprise only of IP addresses then we only need to update the controller list once.
+                if (this.bootstrapServers.stream().allMatch(
+                        inetSocketAddress -> InetAddresses.isInetAddress(inetSocketAddress.getHostString()))) {
+                    final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(this.bootstrapServers.stream()
+                                    .map(address -> new ResolvedServerInfo(
+                                            new InetSocketAddress(address.getHostString(), address.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                    log.info("Updating client with controllers: {}", serverInfoGroup);
+                    this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
+                    started.set(true);
+                    return;
+                }
+                // else case is for controllers with DNS hostnames, which needs to be resolved periodically.
             }
+
+            // Schedule the first discovery immediately.
+            this.scheduledExecutor = Executors.newScheduledThreadPool(1,
+                    new ThreadFactoryBuilder().setNameFormat("fetch-controllers-%d").setDaemon(true).build());
+            this.scheduledFuture = this.scheduledExecutor.schedule(this::getControllers, 0L, TimeUnit.SECONDS);
+            started.set(true);
         }
 
+        // This API should not be under a lock as this is already thread safe and we don't want to block the shutdown.
         @Override
-        @Synchronized
         public void shutdown() {
             Preconditions.checkState(started.compareAndSet(true, false));
             if (this.scheduledExecutor != null) {
                 this.scheduledExecutor.shutdownNow();
             }
         }
 
-        // The controller discovery API invoker.
+        @Override
+        @Synchronized
+        public void refresh() {
+            // Refresh is called as hints when gRPC detects network failures.
+            // We don't want to repeatedly attempt discovery; following logic will limit discovery on failures to
+            // once every FAILURE_RETRY_TIMEOUT_SECONDS seconds. Also we want to trigger discovery sooner on failures.
+            if (started.get()) {
+                if (this.scheduledFuture != null &&
+                        !this.scheduledFuture.isDone() &&
```

Comment:
This is just a heuristic to attempt retry faster (instead of waiting too long), doesn't have to be entirely accurate. Will additionally keep a last update time, this will only enable retries faster in certain corner cases though.